### DISCUSSION TITLE
How to get decoded frames

### DISCUSSION THREAD

**Discussion by stoefln** [NetrisTV/ws-scrcpy#182] [Q&A]
Hi there, thanks for this great project!

I would need the frames decoded in full resolution. I saw that the [h264-converter](https://github.com/xevojapan/h264-converter) is used to decode the data, and forward it to the mediaSource which is assigned to the video tag.
I am not familiar with video decoding at all, so I don't get where I could get the decoded frames from. 
My guess was that it should be somewhere around [here](https://github.com/xevojapan/h264-converter/blob/master/src/index.ts#L152), right? 
Or is there any other place maybe inside of ws-scrcpy where I could access the decompressed frames?

**Comment by drauggres** [comment]
Hi.
h264-converter does not decode video, it creates mp4-container (google muxing) from raw h264-frames and feeds them to a mediaSource. After that, decoding and rendering is done internally by your browser, i.e. you can't access decoded frames (except capturing the content of the video-tag with `.toDataURL()`). 

For decoded frames you need to look at other players/decodes:

https://github.com/NetrisTV/ws-scrcpy/blob/5af62721c450a50172af49c5f29b40898f844163/src/app/player/BroadwayPlayer.ts#L46-L48

https://github.com/NetrisTV/ws-scrcpy/blob/5af62721c450a50172af49c5f29b40898f844163/src/app/player/TinyH264Player.ts#L46-L50

https://github.com/NetrisTV/ws-scrcpy/blob/5af62721c450a50172af49c5f29b40898f844163/src/app/player/WebCodecsPlayer.ts#L89-L92

Also this could be usefull: [Difference between different decoders](https://github.com/NetrisTV/ws-scrcpy/discussions/30)

**Comment by drauggres** [comment]
For broadway.js and tinyh264 in the specified places a `Uint8Array` will contain a decoded frame in [YUV](https://en.wikipedia.org/wiki/YUV)-format. It is later translated into RGB and is rendered on a canvas.

For WebCodecs `frame` is [VideoFrame](https://developer.mozilla.org/en-US/docs/Web/API/VideoFrame) (https://www.w3.org/TR/webcodecs/#videoframe-interface)

**Comment by stoefln** [comment]
@drauggres thanks a lot for the quick reply!

**Comment by ahmedmukhtar1133** [comment]
I would like to read the decoded frames with OpenCv, but it seems OpenCv's imdecode function is not able to read the frame (returns empty object).
Can you tell me more about the format of the frame?
If you'd even know how to transform it to "normal" RGB format, that would be amazing!

**Reply by ahmedmukhtar1133** [reply]
Using frame from `onPictureDecoded` from `BroadWay` player.

**Reply by drauggres** [reply]
> If you'd even know how to transform it to "normal" RGB format, that would be amazing!

https://github.com/NetrisTV/ws-scrcpy/blob/5af62721c450a50172af49c5f29b40898f844163/vendor/h264-live-player/YUVCanvas.ts#L14-L43

**Reply by stoefln** [reply]
@drauggres I'd prefer to use OpenCv since it's much faster.
I checked how to convert YUV to RGB and it seems there are 3 different types

- YUV
- YUV420p
- YUV420sp

However, there are even more conversion types, some of them use postfixes such as "_IYUV" or "_NV12" 
https://docs.opencv.org/3.4/df/d4e/group__imgproc__c.html
<img width="747" alt="image" src="https://user-images.githubusercontent.com/562800/157282787-a6552aed-75d0-4bb0-9dad-a9939fda0068.png">

Any idea which one to use for converting?

The conversion itself would be simple:
https://stackoverflow.com/questions/2231518/how-to-read-a-frame-from-yuv-file-in-opencv#:~:text=To%20convert%20to%20RGB%20from,the%20correct%20conversion%20flag%20enum

**Reply by drauggres** [reply]
Sorry, I am also not an expert in video encoding and don't really know what type is it.
